### PR TITLE
Fix origin position computation

### PR DESCRIPTION
**Pull request by KarelPeeters**
This PR fixes an issue in the computation of the position in the origin file. When a highlighted section starts at the beginning of the line, the origin position used to point to the end of the previous line, instead of the start of the current line. This resulted in outputs like this:

```
error: title
 --> origin.txt:2:4
  |
1 | aaa
2 | bbb
3 | ccc
  | ^^^ annotation
4 | ddd
  |
```

Here `origin.txt:2:4` is wrong, it should be `origin.txt:3:1`, which actually matches the highlighted setting.

As a drive-by fix, 21645ad032dd9117d811db57a5740b6a6406f94a fixes `EndLine`, which was used incorrectly in multiple places that happened to cancel out.

### CODE REVIEW

**Comment by KarelPeeters** [comment]
Ah, I missed those failing test cases, since they're disabled on windows:
https://github.com/rust-lang/annotate-snippets-rs/blob/c84e388949f0e06f4529de060c195e12e6531fbd/tests/fixtures/main.rs#L9-L14
Is there any reason for that, maybe CRLF issues? They seem to behave as expected for me when I just remove that `#[cfg(not(windows))]` line!

The failing test are about cases that try to point one byte past the end of the line. Previously that was indeed possible, but only in very limited situations, eg. the highlighting only worked on EOF cases anyway. See https://github.com/rust-lang/annotate-snippets-rs/issues/108#issuecomment-2290855597 for vaguely relevant discussion about this.

I'm not sure what the best way to proceed here is. Is the "one past the end of the line" use case already adequately supported by just allowing users to point at the newline characters themselves? Ideally there would be some out-of-band way to signal this case, since only using byte indices for this is tricky.

File: src/renderer/display_list.rs [hunk 1]
```diff
@@ -940,9 +940,20 @@ impl<'a> CursorLines<'a> {
...
+
+impl EndLine {
+    /// The number of characters this line ending occupies in bytes.
+    pub(crate) fn len(self) -> usize {
+        match self {
+            EndLine::Eof => 0,
+            EndLine::Lf => 1,
+            EndLine::Crlf => 2,
+        }
+    }
```
Comment (Muscraft) [inline src/renderer/display_list.rs:956]: This should probably be its own PR so it can be evaluated separately

File: tests/formatter.rs [hunk 2]
```diff
@@ -905,3 +905,53 @@ error: unused optional dependency
...
+1 | aaa
+2 | bbb
+3 | ccc
+  |  ^^ annotation
+4 | ddd
+  |
+"#]];
+    let renderer = Renderer::plain();
+    assert_data_eq!(renderer.render(input).to_string(), expected);
+}
```
Comment (Muscraft) [inline tests/formatter.rs:957]: When adding new tests, they should always pass and show the current behavior. Can you please update these to show the current behavior in this commit?

File: src/renderer/display_list.rs [hunk 3]
```diff
@@ -1149,7 +1149,7 @@ fn format_header<'a>(
                 ..
             } = item
             {
-                if main_range >= range.0 && main_range <= range.1 + end_line.len() {
+                if (range.0..range.1 + end_line.len()).contains(&main_range) {
```
Comment (Muscraft) [inline src/renderer/display_list.rs:1152]: This change makes it so highlighting EOF shows the incorrect line-column. I believe the solution to this is to modify the original code to check that `main_range` is less than `range.1 + end_line` not less than or equal, as well as make it so `EOF` adds 1.
```suggestion
                if main_range >= range.0 && main_range < range.1 + max(*end_line as usize, 1) {
```

File: src/renderer/display_list.rs (hunk 1 above)
Comment (KarelPeeters) [reply src/renderer/display_list.rs:956]: Done, split out as #155.

File: tests/formatter.rs (hunk 2 above)
Comment (KarelPeeters) [reply tests/formatter.rs:957]: This is surprising to me, I'm more used to "add failing test, then fix the test in the next commit". Either way is reasonable, I'll change this.

File: src/renderer/display_list.rs (hunk 3 above)
Comment (KarelPeeters) [reply src/renderer/display_list.rs]: That seems to work, thanks for the suggestion!

**Comment by KarelPeeters** [comment]
I think I've addressed all of the comments, and I have rebased this PR on the current master. Let me know if I can do anything else to improve this PR!

**Review by Muscraft — APPROVED** [review]
